### PR TITLE
NativeApp: Enable/disable default hotkeys

### PR DESCRIPTION
**Pull request by snaulX**
Create AppBase::EnableHotkeys() method to enable/disable default hotkeys (such as exit from app on Esc and change fullscreen mode on Shift+Enter in UWP).

### CODE REVIEW

File: NativeApp/src/UWP/App.cpp [hunk 1]
```diff
@@ -285,6 +285,9 @@ void App::OnDisplayContentsInvalidated(DisplayInformation^ sender, Object^ args)
 
 void App::OnKeyDown(Windows::UI::Core::CoreWindow^ sender, Windows::UI::Core::KeyEventArgs^ args)
 {
+    if (!m_Main->EnableHotkeys())
+        return;
```
Comment (TheMostDiligent) [inline NativeApp/src/UWP/App.cpp:289]: This will disable handling the shift key

**Comment by TheMostDiligent** [comment]
This implementation seems to result in incorrect handling of the shift key in UWP. Also, toggling all hot keys on/off at the same time is not very flexible.

I would recommend reworking this change as follows:
- Add `HOT_KEY_FLAGS` flag enum with `HOT_KEY_FLAG_ALLOW_EXIT_ON_ESC` and `HOT_KEY_FLAG_ALLOW_FULL_SCREEN_SWITCH` values.
- Add `HOT_KEY_FLAGS GetHotKeyFlags()` method instead of `EnableHotkeys()`

**Comment by snaulX** [comment]
Hmm.. Nice idea. Maybe also add `HOT_KEY_FLAG_NONE = 0` (or `HOT_KEY_FLAG_ALLOW_NONE`)?

File: NativeApp/include/AppBase.hpp [hunk 2]
```diff
@@ -70,6 +70,19 @@ class AppBase
...
+    /// Flags to handle default hotkeys
+    enum HOT_KEY_FLAGS : Uint32
+    {
+        /// App doesn't react on default hotkeys
+        HOT_KEY_FLAG_NONE = 0,
+
+        /// App will exit by pressing Escape
+        HOT_KEY_FLAG_ALLOW_EXIT_ON_ESC = 1,
+
+        /// App wil change fullscreen mode on Shift+Enter (only in UWP)
```
Comment (TheMostDiligent) [inline NativeApp/include/AppBase.hpp:82]: Typo: wil

File: NativeApp/include/AppBase.hpp [hunk 3]
```diff
@@ -70,6 +70,19 @@ class AppBase
...
     };
 
+    /// Flags to handle default hotkeys
+    enum HOT_KEY_FLAGS : Uint32
+    {
+        /// App doesn't react on default hotkeys
+        HOT_KEY_FLAG_NONE = 0,
+
+        /// App will exit by pressing Escape
+        HOT_KEY_FLAG_ALLOW_EXIT_ON_ESC = 1,
```
Comment (TheMostDiligent) [inline NativeApp/include/AppBase.hpp:80]: Please use `1 << 0` for consistency

File: NativeApp/include/AppBase.hpp [hunk 4]
```diff
@@ -70,6 +70,19 @@ class AppBase
...
+
+        /// App wil change fullscreen mode on Shift+Enter (only in UWP)
+        HOT_KEY_FLAG_ALLOW_FULL_SCREEN_SWITCH = 1 << 1
+    };
```
Comment (TheMostDiligent) [inline NativeApp/include/AppBase.hpp:84]: Please add the `HOT_KEY_FLAG_ALL = ~0u` value.

Please add the `DEFINE_FLAG_ENUM_OPERATORS` macro, see example [here](DEFINE_FLAG_ENUM_OPERATORS).

File: NativeApp/include/AppBase.hpp [hunk 5]
```diff
@@ -156,6 +169,12 @@ class AppBase
     {
         return false;
     }
+
+    /// Returns default hotkeys handling flags
+    virtual HOT_KEY_FLAGS GetHotKeyFlags() const
+    {
+        return HOT_KEY_FLAG_ALLOW_EXIT_ON_ESC | HOT_KEY_FLAG_ALLOW_FULL_SCREEN_SWITCH;
```
Comment (TheMostDiligent) [inline NativeApp/include/AppBase.hpp:176]: Please return the `HOT_KEY_FLAG_ALL` here. If new enum values are add, it is almost guaranteed that this function will not be updated.

File: NativeApp/src/Linux/LinuxMain.cpp [hunk 6]
```diff
@@ -529,7 +529,7 @@ int x_main(int argc, const char* const* argv)
             }
         }
 
-        if (EscPressed)
+        if (EscPressed && TheApp->GetHotKeyFlags() & HOT_KEY_FLAG_ALLOW_EXIT_ON_ESC)
```
Comment (TheMostDiligent) [inline NativeApp/src/Linux/LinuxMain.cpp:532]: Please use parens to group bitwise &

File: NativeApp/src/Win32/WinMain.cpp [hunk 7]
```diff
@@ -196,7 +196,7 @@ LRESULT CALLBACK MessageProc(HWND wnd, UINT message, WPARAM wParam, LPARAM lPara
             return 0;
 
         case WM_CHAR:
-            if (wParam == VK_ESCAPE)
+            if (wParam == VK_ESCAPE && g_pTheApp->GetHotKeyFlags() & HOT_KEY_FLAG_ALLOW_EXIT_ON_ESC)
```
Comment (TheMostDiligent) [inline NativeApp/src/Win32/WinMain.cpp:199]: Please use parens to group bitwise &

**Review by TheMostDiligent — CHANGES_REQUESTED** [review]

**Pull request merged**